File: android/src/main/java/com/media/module/appstate/AppStateNativeManagerModule.java
package com.media.module.appstate;

import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;

public class AppStateNativeManagerModule extends ReactContextBaseJavaModule {

    // =============================================================================================
    // ATTRIBUTES ==================================================================================

    private static final String TAG = "AppStateNative";

    // =============================================================================================
    // CONSTRUCTOR =================================================================================

    public AppStateNativeManagerModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    // =============================================================================================
    // METHODS =====================================================================================

    @Override
    public String getName() {
        return "AppStateNativeManagerModule";
    }

    // =============================================================================================
    // EVENT =======================================================================================

    // =============================================================================================
    // CLASS =======================================================================================

    public static final class Event {
        public static final String ON_RESUME = "onResume";
        public static final String ON_ACTIVE = "onActive";
        public static final String ON_LOST_FOCUS = "onLostFocus";
        public static final String ON_PAUSE = "onPause";
        public static final String ON_DESTROY = "onDestroy";
    }

    public static final class State {
        public static final int RESUME = 0;
        public static final int ACTIVE = 1;
        public static final int LOST_FOCUS = 2;
        public static final int PAUSE = 3;
        public static final int STOP = 4;
    }
}

File: ios/AppStateNativeManager/AppStateNativeManagerModule.swift
//
//  AppStateNativeManagerModule.swift
//  RNReactNativeMedia
//
//  Created by Haroldo Shigueaki Teruya on 14/08/2018.
//  Copyright © 2018 Facebook. All rights reserved.
//

import Foundation
import AVFoundation

@objc(AppStateNativeManagerModule)
class AppStateNativeManagerModule: NSObject {

    let TAG: String = "AppStateNative"
    var isListening = false

    // =============================================================================================
    // ATRIBUTES ===================================================================================

    struct Event {
        static let ON_ACTIVE = "onActive";
        static let ON_PAUSE = "onPause";
        static let ON_STOP = "onStop";
        static let ON_DESTROY = "onDestroy";
    }

    var bridge: RCTBridge!

    // =============================================================================================
    // CONSTRUCTOR =================================================================================

    // =============================================================================================
    // METHODS =====================================================================================

    @objc func addAllListener() {

        if !isListening {
            NotificationCenter.default.addObserver(
                self,
                selector: #selector(self.onHostPause),
                name: .UIApplicationWillResignActive,
                object: nil)

            NotificationCenter.default.addObserver(
                self,
                selector: #selector(self.onHostActive),
                name: .UIApplicationDidBecomeActive,
                object: nil)

            NotificationCenter.default.addObserver(
                self,
                selector: #selector(self.onHostStop),
                name: .UIApplicationDidEnterBackground,
                object: nil)

            NotificationCenter.default.addObserver(
                self,
                selector: #selector(self.onHostDestroy),
                name: .UIApplicationWillTerminate,
                object: nil)

            isListening = true
        }
    }

    @objc func onHostPause() {
        self.emitEvent(eventName: Event.ON_PAUSE, data: nil)
    }

    @objc func onHostActive() {
        self.emitEvent(eventName: Event.ON_ACTIVE, data: nil)
    }

    @objc func onHostStop() {
        self.emitEvent(eventName: Event.ON_STOP, data: nil)
    }

    @objc func onHostDestroy() {
        self.emitEvent(eventName: Event.ON_DESTROY, data: nil)
    }

    func emitEvent(eventName: String, data: Any?) -> Void {
        NSLog("AppDelegate AppState emitEvent \(eventName)")
        if self.bridge != nil, self.bridge.eventDispatcher() != nil {
            self.bridge.eventDispatcher().sendAppEvent(withName: eventName, body: data)
        }
    }
}
